Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm

```diff
@@ -280,81 +289,71 @@
                 RELEASE_LOG_ERROR(MediaStream, "AudioSampleBufferCompressor getting kAudioConverterCompressionMagicCookie failed with %d", error);
-                return nil;
-            }
         }
 
         CMFormatDescriptionRef destinationFormatDescription;
-        if (auto error = PAL::CMAudioFormatDescriptionCreate(kCFAllocatorDefault, &m_destinationFormat, 0, NULL, cookieSize, cookie.data(), NULL, &destinationFormatDescription)) {
+        if (int error = PAL::CMAudioFormatDescriptionCreate(kCFAllocatorDefault, &m_destinationFormat, 0, nullptr, cookieSize, cookie.data(), nullptr, &destinationFormatDescription)) {
             RELEASE_LOG_ERROR(MediaStream, "AudioSampleBufferCompressor CMAudioFormatDescriptionCreate failed with %d", error);
-            return nil;
+            return makeUnexpected(error);
         }
         m_destinationFormatDescription = adoptCF(destinationFormatDescription);
         m_gdrCountNum = gradualDecoderRefreshCount();
     }
 
-    char *data = static_cast<char*>(fillBufferList.mBuffers[0].mData);
-    size_t dataSize = fillBufferList.mBuffers[0].mDataByteSize;
-
-    CMBlockBufferRef blockBuffer;
-    if (auto error = PAL::CMBlockBufferCreateWithMemoryBlock(kCFAllocatorDefault, NULL, dataSize, kCFAllocatorDefault, NULL, 0, dataSize, kCMBlockBufferAssureMemoryNowFlag, &blockBuffer)) {
-        RELEASE_LOG_ERROR(MediaStream, "AudioSampleBufferCompressor CMBlockBufferCreateWithMemoryBlock failed with %d", error);
-        return nil;
-    }
-    auto buffer = adoptCF(blockBuffer);
-
-    if (auto error = PAL::CMBlockBufferReplaceDataBytes(data, buffer.get(), 0, dataSize)) {
-        RELEASE_LOG_ERROR(MediaStream, "AudioSampleBufferCompressor CMBlockBufferReplaceDataBytes failed with %d", error);
-        return nil;
-    }
-
     CMSampleBufferRef rawSampleBuffer;
-    auto error = PAL::CMAudioSampleBufferCreateWithPacketDescriptions(kCFAllocatorDefault, buffer.get(), true, NULL, NULL, m_destinationFormatDescription.get(), 1, m_currentNativePresentationTimeStamp, &m_destinationPacketDescriptions, &rawSampleBuffer);
-    if (error) {
+    if (int error = PAL::CMAudioSampleBufferCreateWithPacketDescriptions(kCFAllocatorDefault, nullptr, false, nullptr, nullptr, m_destinationFormatDescription.get(), numSamples, m_currentNativePresentationTimeStamp, isPCM() ? nullptr : m_destinationPacketDescriptions.data(), &rawSampleBuffer)) {
         RELEASE_LOG_ERROR(MediaStream, "AudioSampleBufferCompressor CMAudioSampleBufferCreateWithPacketDescriptions failed with %d", error);
-        return nil;
+        return makeUnexpected(error);
     }
-
     auto sampleBuffer = adoptCF(rawSampleBuffer);
+    if (int error = PAL::CMSampleBufferSetDataBufferFromAudioBufferList(sampleBuffer.get(), kCFAllocatorDefault, kCFAllocatorDefault, kCMSampleBufferFlag_AudioBufferList_Assure16ByteAlignment, fillBufferList.list())) {
```

Comment:
I wonder if we have elsewhere this WebAudioBufferList -> CMSampleBuffer conversion. Maybe it could be moved to a utility routine somewhere if that is the case.

Reply:
I was thinking the same thing. I'll do that in a follow-up. It's a very common way to go about it ever since I discovered `CMSampleBufferSetDataBufferFromAudioBufferList` :)